src/Units/Unit.java

```diff
@@ -83,39 +82,61 @@ else if( CO.money >= proportionalCost )
   }
 
   /**
-   * @return the base damage this unit would do against the specified target,
-   * at the specified range, based on whether it will move before striking
-   * Chooses the first weapon on the list that can deal damage.
+   * @return whether or not this unit can attack the given unit type at the
+   * specified range, accounting for the possibility of moving first.
    */
-  public double getBaseDamage(UnitModel target, int range, boolean moved)
+  public boolean canAttack(UnitModel targetType, int range, boolean afterMoving)
   {
     // if we have no weapons, we can't hurt things
     if( weapons == null )
-      return 0;
-    double maxDamage = 0;
+      return false;
+
+    boolean canHit = false;
     for( int i = 0; i < weapons.length; i++ )
     {
-      // If the weapon isn't mobile, we shouldn't be able to fire if we moved.
-      if( moved && !weapons[i].model.canFireAfterMoving )
+      if( afterMoving && !weapons[i].model.canFireAfterMoving )
       {
+        // If we are planning to move first, and the weapon
+        // can't shoot after moving, then move along.
         continue;
       }
-      double currentDamage = weapons[i].getDamage(target, range);
-      if( maxDamage < currentDamage )
+      if( weapons[i].getDamage(targetType, range) > 0 )
       {
-        maxDamage = currentDamage;
-        currentWeapon = weapons[i];
+        canHit = true;
+        break;
       }
     }
-    return maxDamage;
+    return canHit;
   }
 
-  // for the purpose of letting the unit know it has attacked.
-  public void fire(final Unit defender)
+  /**
+   * Select the weapon owned by this unit that can inflict the
+   * most damage against the chosen target
+   * @param target
+   * @param range
+   * @param moved
+   * @return
+   */
+  public Weapon chooseWeapon(UnitModel targetType, int range, boolean moved)
```

Comment:
Should probably unify nomenclature with canAttack - make the boolean "afterMoving"